src/monitor.py

```diff
@@ -115,3 +119,147 @@ def read_recent_logs(self, job: JobDto, tail: int = 20) -> str:
+        # A "check" return True on deployment success, False on deployment failure, and None if unsure.
+        # The timeouts, number of checks, and loop_sleeps are mostly based on heuristics and vibes:
+        #   We don't want to spam the cluster with requests
+        #   But we also do not want the deployment to take much longer
+        # If the deployment just works, it will be delayed by O(n_consecutive_checks * total_loop_time)
+        # TODO: If we are really confident in 'kubectl rollout status' we could spawn it in a separate thread and wait for it to finish with a watch='true'
+        n_consecutive_checks = 3
+        checks: list[None | bool] = [None] * n_consecutive_checks
+        loop_sleep_time = 8
+        start_time = time.time()
+        timeout = start_time + 10 * 60 # (n_minutes * seconds/minute) - we experienced a deploy failing after 8 minutes.
+        i = 0
+        now = time.time()
+        while now < timeout:
+            is_rollout_okay, rollout_status = self.check_rollout_status(resource_name)
+            is_event_okay, events = self.check_k8s_events_for_errors(resource_name)
+            checks[i] = self.combine_checks(is_rollout_okay, is_event_okay)
+
+            i = (i + 1) % n_consecutive_checks
+            now = time.time()
+
+            if None in checks:
+                time.sleep(loop_sleep_time)
+                continue
+
+            if all(checks):
+                return
+
+            if not any(checks):
+                error_msg = f'{n_consecutive_checks} consecutive early errors detected in {resource_name}. Rollout status: {rollout_status}. Kubernetes events: {events}'
+                raise RuntimeError(error_msg)
+            time.sleep(loop_sleep_time)
+
+
+    def combine_checks(self, is_rollout_okay: bool | None, is_event_okay: bool | None) -> bool | None:
+        # If is_rollout_okay is None, then we trust is_event_okay instead
+        # If is_event_okay is None, then we trust is_rollout_okay instead
+        # If neither are None, we need both of them to say that the deployment went well
+        if is_event_okay is None:
+            return is_rollout_okay
+        if is_rollout_okay is None:
+            return is_event_okay
+        return is_rollout_okay and is_event_okay
+
+
+    def check_rollout_status(self, resource_name: str, watch: str = 'false') -> tuple[bool | None, str]:
+        # Some possible outputs:
+        # https://github.com/kubernetes/kubernetes/blob/f1d63237edf908aae577d3da60276151c18ffee0/staging/src/k8s.io/kubectl/pkg/polymorphichelpers/rollout_status.go#L89
+        successful_rollouts_messages =[
+            f'deployment \"{resource_name}\" successfully rolled out',
+            ]
+        failure_rollouts_messages = [
+            f'deployment \"{resource_name}\" exceeded its progress deadline'
+        ]
+
+        cmd = f'kubectl rollout status deployment/{resource_name} --watch={watch} --namespace {K8S_NAMESPACE} --revision=0'
```

Comment:
I have the "watch" there in case we wanted to run it in another thread.

revision=0 is always the latest, but I see that is default anyway - I'll remove it.